### DISCUSSION TITLE
Empty body schema when adding file upload ability to endpoint

### DISCUSSION THREAD

**Discussion by psowa001** [fastapi/fastapi#9023] [Questions] [answered]
I am developing endpoint on which users will be able to create articles.
In the body there will be information like description, body, title, etc.
It was working perfectly but when I added ability to add image to article in swagger there is empty request body schema for article, like on below image.

![Capture](https://user-images.githubusercontent.com/37558184/111454057-85463a80-8714-11eb-9464-be357bf5fb9f.PNG)

Code:
```
# ARTICLE
class ArticleBase(BaseModel):
    title: str
    description: str
    content: str
    date_time: datetime
    public: bool

class ArticleCreate(ArticleBase):
    pass

class Article(ArticleBase):
    id: int
    image: str
    author_id: int

@router.post("/articles/")
async def create_user_article(article: ArticleCreate, image: Optional[bytes] = File(None), token: schemas.TokenData = Depends(utils.check_token)):
    return await crud.create_user_article(user_id=token.user_id, article=article)
```

Any ideas?

**Comment by falkben** [comment]
what is `schemas`?

**Comment by krishnardt** [comment]
The pydantic schemas you might be using is not correct.. and also there is an issue in the functional parameters i.e., usage of image.

**Comment by psowa001** [comment]
Hi @falkben and @krishnardt, thank you for your answers. Please look at my first post again, I updated it.
@krishnardt could you please indicate there is the issue in my code?
Everything is displaying correctly when remove
```
image: Optional[bytes] = File(None)
```
I want uploading image to be optional.

**Comment by cmflynn** [comment]
I believe this is why(https://fastapi.tiangolo.com/tutorial/request-files/): 

You can declare multiple File and Form parameters in a path operation, but you can't also declare Body fields that you expect to receive as JSON, as the request will have the body encoded using multipart/form-data instead of application/json.

This is not a limitation of FastAPI, it's part of the HTTP protocol.

**Comment by krishnardt** [comment]
Can you try replacing image with the following code....

`File(None) --> File(...)`

**Comment by juntatalor** [comment, marked as answer]
I think that it is not possible to use json body and file (image) upload in one handler. Because json is parsed as the whole request body (application/json) and file upload is implemented via multipart/formdata (also whole request body in different format)

I can suggest three possible solutions:
1) Use different handler to attach images (prefered)
2) Use a str field in ArticleCreate model and pass image with base64 encoding
3) Use multipart/formdata for the whole request (article data + files: https://fastapi.tiangolo.com/tutorial/request-forms-and-files/)

**Comment by dotX12** [comment]
@psowa001, @juntatalor 
Better to know the answer late than never :)
A small refinement that allows you to disassemble the models correctly.

https://github.com/dotX12/pyfa-converter


```python3
from datetime import datetime
from typing import Optional

from fastapi import FastAPI, UploadFile, File, Form
from pydantic import BaseModel, Field

from pyfa_converter import FormDepends, PyFaDepends

app = FastAPI()


class PostContractBodySchema(BaseModel):
    title: str = Field(..., description="Description title")
    date: Optional[datetime] = Field(
        None, description="Example: 2021-12-14T09:56:31.056Z"
    )


@app.post("/form-data-body")
async def example_foo_body_handler(
    data: PostContractBodySchema = FormDepends(PostContractBodySchema),
    # data1: PostContractBodySchema = PyFaDepends( # OR
    #         model= PostContractBodySchema, _type=Form
    #     ),
    document: UploadFile = File(...),
):
    return {"title": data.title, "date": data.date, "file_name": document.filename}

```
